include/base_interface.h

```diff
@@ -0,0 +1,280 @@
+   */
+  BaseInterface(const unsigned int &n_comp,
+                const unsigned int &n_matrices,
+                const std::string &name="",
+                const std::string &default_fe="FE_Q(1)",
+                const std::string &default_component_names="u",
+                const std::string &default_differential_components="");
+
+  void init ();
+
+
+  virtual void declare_parameters (ParameterHandler &prm);
+
+  virtual void parse_parameters_call_back() {};
+
+  /**
+   * returns the vector @p of differential blocks
+   */
+  const std::vector<unsigned int> get_differential_blocks() const;
+
+  /**
+   * This function is used to modify triangulation using boundary_id or manifold_id.
+   * In the case a signal is required, this is the function to modify.
+   */
+  virtual void postprocess_newly_created_triangulation(Triangulation<dim, spacedim> &tria) const;
+
+  /**
+   * Initialize all data required for the system
+   *
+   * This function is used to initialize the internal variables
+   * according to the given arguments, which are
+   * @p solution, @p solution_dot, @p t and @p alpha.
+   */
+  virtual void initialize_data(const typename LAC::VectorType &solution,
+                               const typename LAC::VectorType &solution_dot,
+                               const double t,
+                               const double alpha) const;
+
+
+  /**
+   * Definition of energies and residuals
+   */
+  virtual void assemble_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &,
+                                               FEValuesCache<dim,spacedim> &,
+                                               std::vector<Sdouble> &energies,
+                                               std::vector<std::vector<double> > &local_residuals,
+                                               bool compute_only_system_terms) const;
+
+  /**
+   * Definition of energies and residuals
+   */
+  virtual void assemble_energies_and_residuals(const typename DoFHandler<dim,spacedim>::active_cell_iterator &,
+                                               FEValuesCache<dim,spacedim> &,
+                                               std::vector<SSdouble> &energies,
+                                               std::vector<std::vector<Sdouble> > &local_residuals,
+                                               bool compute_only_system_terms) const;
+
+  /**
+   * This function can be overloaded to directly implement the local
+   * matrices (i.e. as it is usally done in standard FE codes)
+   */
+  virtual void assemble_local_matrices (const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
+                                        FEValuesCache<dim,spacedim> &scratch,
+                                        CopyData &data) const;
+
+  virtual void assemble_local_system_residual (const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
```

Comment:
Assemble the local system residual associated to the given cell. 

This function is called to evaluate the local system residual at each Newton iteration.